cmd/kubeadm/app/cmd/phases/kubeconfig.go

```diff
@@ -19,77 +19,150 @@ package phases
+	KubeConfigDir() string
+}
 
-	kubeletKubeconfigLongDesc = fmt.Sprintf(normalizer.LongDesc(`
-		Generates the kubeconfig file for the kubelet to use and saves it to %s file.
+// NewKubeConfigPhase creates a kubeadm workflow phase that creates all kubeconfig files necessary to establish the control plane and the admin kubeconfig file.
+func NewKubeConfigPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "kubeconfig",
+		Short: "Generates all kubeconfig files necessary to establish the control plane and the admin kubeconfig file",
+		Phases: []workflow.Phase{
+			NewKubeConfigFilePhase(kubeadmconstants.AdminKubeConfigFileName),
+			NewKubeConfigFilePhase(kubeadmconstants.KubeletKubeConfigFileName),
+			NewKubeConfigFilePhase(kubeadmconstants.ControllerManagerKubeConfigFileName),
+			NewKubeConfigFilePhase(kubeadmconstants.SchedulerKubeConfigFileName),
+		},
+		Run: runKubeConfig,
+	}
+}
 
-		Please note that this should *only* be used for bootstrapping purposes. After your control plane is up,
-		you should request all kubelet credentials from the CSR API.
-		`+cmdutil.AlphaDisclaimer), filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.KubeletKubeConfigFileName))
+// NewKubeConfigFilePhase creates a kubeadm workflow phase that creates a kubeconfig file.
+func NewKubeConfigFilePhase(kubeConfigFileName string) workflow.Phase {
+	return workflow.Phase{
+		Name:  kubeconfigFilePhaseProperties[kubeConfigFileName].name,
+		Short: kubeconfigFilePhaseProperties[kubeConfigFileName].short,
+		Long:  fmt.Sprintf(kubeconfigFilePhaseProperties[kubeConfigFileName].long, kubeConfigFileName),
+		Run:   runKubeConfigFile(kubeConfigFileName),
+	}
+}
 
-	controllerManagerKubeconfigLongDesc = fmt.Sprintf(normalizer.LongDesc(`
-		Generates the kubeconfig file for the controller manager to use and saves it to %s file.
-		`+cmdutil.AlphaDisclaimer), filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ControllerManagerKubeConfigFileName))
+func runKubeConfig(c workflow.RunData) error {
+	data, ok := c.(kubeConfigData)
+	if !ok {
+		return errors.New("kubeconfigFile phase invoked with an invalid data struct")
+	}
 
-	schedulerKubeconfigLongDesc = fmt.Sprintf(normalizer.LongDesc(`
-		Generates the kubeconfig file for the scheduler to use and saves it to %s file.
-		`+cmdutil.AlphaDisclaimer), filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.SchedulerKubeConfigFileName))
+	fmt.Printf("[kubeconfig] Using kubeconfig folder %q\n", data.KubeConfigDir())
+	return nil
+}
 
-	userKubeconfigLongDesc = normalizer.LongDesc(`
-		Outputs a kubeconfig file for an additional user.
-		` + cmdutil.AlphaDisclaimer)
+// runKubeConfigFile executes kubeconfig creation logic.
+func runKubeConfigFile(kubeConfigFileName string) func(workflow.RunData) error {
+	return func(c workflow.RunData) error {
+		data, ok := c.(kubeConfigData)
+		if !ok {
+			return errors.New("kubeconfigFile phase invoked with an invalid data struct")
```

Comment:
perhaps use `kubeconfig`, as `kubeconfigFile` is not a real phase name?
optionally use `kubeConfigFileName` to clarify?

Reply:
I think having the print [phase] = error [phase] should probably match.